expr/func_movingwindow.go

```diff
@@ -0,0 +1,154 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/schema"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average"}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncMovingWindow) Context(context Context) Context {
+	var err error
+	s.shiftOffset, err = s.getWindowSeconds(context)
+	if err != nil {
+		// panic? - windowSize specified as number of points
```

Comment:
I think we need to extend either the `Context` type to hold an `error` or the function to return one (which would require a lot of function modifications).

Reply:
Sure. I'll follow your lead on timeShift. let me know what the recommended way is.

Reply:
timeShift doesn't really do it. The native timeShift supports all options. There is a small corner case around time 0, but that's not really something that is likely to happen, so it's just a TODO

Congrats, you are the first :)